Add Django Util With Decorator and Views

waprin opened the pull request:
This Django extension adds the views, decorators, and signals to oauth2client.

This is heavily based on the work @jonparrott did for oauth2client/flask_util.

Currently the only storage offered uses Django sessions. Based on issue #319 more storage options are planned along with integration to the existing CredentialsField in django_orm. 

This adds an environment variable to the tox script. This might not be strictly necessary, but it did simplify the tests a lot, since Django settings get looked up at import time and cached so it's easiest if they have an environment variable pointing to an existing settings module when the tests are started.

This could have been squeezed into a single file, but  I spilt it into separate files based on existing Django extension conventions.

Commit:
Add Django Util With Decorator

Adds a submodule that provides views, decorator, and signals to help
a Django web application complete OAuth2 authorization.

dhermes commented on tests/test_django_orm.py:
```diff
@@ -30,10 +30,11 @@
-sys.modules['django_settings'] = django_settings = imp.new_module(
-    'django_settings')
-django_settings.SECRET_KEY = 'xyzzy'
+os.environ['DJANGO_SETTINGS_MODULE'] = 'tests.test_django_settings'
```
Thanks for this

dhermes commented:
All new files added should have 100% coverage (towards #212).

From:
https://coveralls.io/builds/3925551

The following have less than 100%:
- `django_util/apps.py`: 0%
- `django_util/storage.py`: 89.47%
- `tests/test_django_util.py`: 98.05% (every line that isn't run should be removed or `pragma: NO COVER`ed)

dhermes commented:
Running `tox -e cover` locally it also looks like

```
oauth2client/django_util/storage.py         19      2      4      0    83%   51-52
oauth2client/django_util/views.py           62      0     10      1    99%   133->135
```

dhermes commented:
If running the tests create `db.sqlite3`, you should add it to the `.gitignore`.

waprin commented:
@dhermes thanks for taking a look, I now I have 100% coverage on all the main files and fixe the gitignore. The test file is still missing 3 lines, I added #pragma: no cover, since they are test views and the point is that the decorator redirects before they are run. I could maybe mock them but I think it's clearer to have them as functions with the decorator...but for some reason coverage isn't respecting the pragma. Investigating why...

dhermes commented:
Check out [`.coveragerc`][1] the `NO COVER` has to be uppercase. (I just did `git grep -i pragma` to get an idea.)

[1]: https://github.com/google/oauth2client/blob/311a53fbcae550526a74057f27cb9df73a0af009/.coveragerc#L5

waprin commented:
Wow, thanks for the headsup on the pragma case sensitivity, it was really confusing me, forgot to check all the hidden files! Ok, now I should have 100% test coverage. I also added NO COVER to apps.py as well, it's covered in 2.7+ but it skips the code in 2.6, and since the 2.6 case isn't run by coverage the 'else' branch isn't covered. It's basically just a metadata class for recent Django versions, which dropped support for 2.6 and so the version of Django tox installs for 2.6 doesn't have the imports.

So I think this is ready for further review or merge whenever you or nathaniel find the time. jonparrot already did a review.

jonparrott commented:
Summoning @anthmgoogle, he should give this a review as well.

dhermes commented:
Sorry, I don't plan on reviewing (not enough cycles). I was just "protecting" my work towards #212 by making sure big changes were fully tested.

I actually took a stab at testing `django_orm.py` and stopped at the point that I'd need a test DB. Now that you've added one, maybe you could easily cover the rest of `django_orm.py` (just the `Storage` class) and get us one step closer to 100% coverage?

nathanielmanistaatgoogle commented:
I am able to review this but it would be best if I were last. Poke me after both @jonparrott and @anthmgoogle are satisfied?

waprin commented:
@dhermes Will do, thanks for the help. I planned to do that in a separate PR as part of Storage improvement efforts but I could add it onto this one. Also curious if it would make sense to move django_orm into this package or if that's not ok since it breaks existing code (I'm guessing it's ok since this library has releases). 

@nathanielmanistaatgoogle sounds good to me, @jonparrott already took a look (and it's a lot of extremely similar code to his own) but I'll hold off for @anthmgoogle and ping you after.

jonparrott commented:
I'm satisfied in terms of overall approach, parity with `flask_util`, documentation, and style (though I'm not nearly as thorough as Nathaniel). @anthmngoogle to confirm it fits with his expectations to satisfy #223.

dhermes commented:
@waprin You should do it in a separate PR. The smaller a PR is, the easier it is on everyone. As for moving `django_orm`, it seems like the right thing to do. An alias could be set up and deprecated slowly over time (@jonparrott has ideas to this effect).

anthmgoogle commented on oauth2client/django_util/__init__.py:
```diff
@@ -0,0 +1,167 @@
+
+
+    @required
+    def requires_default_scopes(request):
```
Most implementations have a "required" and an "optional" mode where you can programatically check if you have authorizaiton. The optional version may also be good to show in this top level example, since it is fairly common in practice. I could not locate such a feature later on. It may be worth adding or creating an issue to add this later.

anthmgoogle commented on oauth2client/django_util/storage.py:
```diff
@@ -0,0 +1,52 @@
+
+
+def get_storage(request):
+    # TODO: Make this pluggable with different storage providers
```
Might be better to add a name for the todo or create an issue track this.

nathanielmanistaatgoogle replied:
I see you've already created an issue - please format this as "TODO(issue 319): <text>".

waprin replied:
done in incoming push

anthmgoogle commented on oauth2client/django_util/views.py:
```diff
@@ -0,0 +1,137 @@
+        server_csrf = request.session[_CSRF_KEY]
+        code = request.GET['code']
+    except KeyError:
+        return HttpResponseBadRequest("Invalid Request")
```
We could perhaps help with a more helpful message. This can happen in some cases where sessions get stale, Perhaps something like "Invalid authorization session, please retry authorization".

anthmgoogle commented on oauth2client/django_util/views.py:
```diff
@@ -0,0 +1,137 @@
+        client_csrf = state['csrf_token']
+        return_url = state['return_url']
+    except (ValueError, KeyError):
+        return HttpResponseBadRequest('Invalid request state')
```
These "invalid request state" errors seem unlikely unless we have errors in the components. Perhaps add slightly more specific message  in case they occur without a stack trace to point to the line.

anthmgoogle commented:
Finished a review pass. I suggested one follow up feature that could be done at a later time. Some minor suggestions about error messages. 

Looking great. Really exciting to see this functionality.

waprin commented:
@anthmgoogle Thanks for review, sorry for delay in turnaround. Let me know if you can +1 now and I can pass it along to Nathaniel.

I agree that missing the optional oauth was a major oversight so I added it. Now there are two decorators `oauth_required` and `oauth_enabled`, with similar semantics to the webapp2 and flask decorators.  The biggest difference is they attach their instance to the Django request object. 

Also based on your comments, I also improved the error messages in the oauth2callback, and reference the Github issue for the pluggable storage feature in the TODO. That is next up on our radar along with some other refactoring.

anthmgoogle commented:
LGTM.

waprin commented:
@nathanielmanistaatgoogle ready for your pass when you have a chance.

nathanielmanistaatgoogle commented:
I am oversubscribed for the next few weeks but happy to accommodate if necessary; how time-sensitive is this?

jonparrott commented:
I suppose that's up to @anthmgoogle.

I don't want to start on #326, #319, and #225 until this is in, but those are rather long-tail issues.

nathanielmanistaatgoogle commented on oauth2client/django_util/__init__.py:
```diff
@@ -0,0 +1,284 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
```
One blank line between license header block and module doc string.

waprin replied:
done in incoming push

nathanielmanistaatgoogle commented on oauth2client/django_util/__init__.py:
```diff
@@ -0,0 +1,284 @@
+
+   oauth2_authorized.connect(test_callback)
+
+"""
```
One blank line between module doc string and imports.

waprin replied:
done in incoming push

nathanielmanistaatgoogle commented on oauth2client/django_util/__init__.py:
```diff
@@ -0,0 +1,284 @@
+import httplib2
+
+from django.conf import settings
+from django.core.exceptions import ImproperlyConfigured
```
Import only modules, not classes, functions, or other elements of modules: https://google.github.io/styleguide/pyguide.html#Imports.

waprin replied:
done in incoming push (though not a huge fan of this rule in all the cases it applies)

nathanielmanistaatgoogle commented on oauth2client/django_util/__init__.py:
```diff
@@ -0,0 +1,284 @@
+    if getattr(settings, 'GOOGLE_OAUTH2_CLIENT_SECRETS_JSON', None):
+        return _load_client_secrets(
+            settings_instance.GOOGLE_OAUTH2_CLIENT_SECRETS_JSON)
+    elif getattr(settings_instance, "GOOGLE_OAUTH2_CLIENT_ID", None) and \
```
"Do not use backslash line continuation.": https://google.github.io/styleguide/pyguide.html?showone=Line_length#Line_length.

waprin replied:
done in incoming commit

nathanielmanistaatgoogle commented on oauth2client/django_util/__init__.py:
```diff
@@ -0,0 +1,284 @@
+        return settings_instance.GOOGLE_OAUTH2_CLIENT_ID, \
+            settings_instance.GOOGLE_OAUTH2_CLIENT_SECRET
+    else:
+        raise ImproperlyConfigured("Must specify either "
```
Line-wrap immediately after the opening parenthesis and you can state this string literal in fewer lines.

waprin replied:
good idea, done in incoming commit

nathanielmanistaatgoogle commented on oauth2client/django_util/__init__.py:
```diff
@@ -0,0 +1,284 @@
+                                   "settings.py")
+
+
+class OAuth2Settings(object):
```
This class defines no public methods - what use is it? If it is intended to be used for its attributes, add an "Attributes:" section to its doc string. (Example available at https://google.github.io/styleguide/pyguide.html?showone=Comments#Comments.)

waprin replied:
done in incoming commit

nathanielmanistaatgoogle commented on oauth2client/django_util/__init__.py:
```diff
@@ -0,0 +1,284 @@
+
+
+class UserOAuth2(object):
+    """ Class to create oauth2 objects on Django request objects containing
```
Spurious whitespace between triple-quote and text.

waprin replied:
done in incoming commit

nathanielmanistaatgoogle commented on oauth2client/django_util/__init__.py:
```diff
@@ -0,0 +1,284 @@
+    """
+
+    def __init__(self, request, scopes=None, return_url=None):
+        """ Initialize the Oauth2 Object
```
Spurious whitespace between triple-quote and text.

waprin replied:
done in incoming commit

nathanielmanistaatgoogle commented on oauth2client/django_util/__init__.py:
```diff
@@ -0,0 +1,284 @@
+        return None
+
+
+def _redirect_with_params(url_name, *args, **kwargs):
```
Please place this function definition above (earlier in the file text) all of its use-sites.

waprin replied:
done in incoming commit

nathanielmanistaatgoogle commented on oauth2client/django_util/__init__.py:
```diff
@@ -0,0 +1,284 @@
+
+
+def _redirect_with_params(url_name, *args, **kwargs):
+    """ Helper method to create a redirect response that uses GET URL parameters"""
```
Spurious doc string space.

waprin replied:
done in incoming commit

nathanielmanistaatgoogle commented on oauth2client/django_util/__init__.py:
```diff
@@ -0,0 +1,284 @@
+
+def _redirect_with_params(url_name, *args, **kwargs):
+    """ Helper method to create a redirect response that uses GET URL parameters"""
+    from six.moves.urllib.parse import urlencode
```
What's up with an import not at top-level?

waprin replied:
done in incoming commit

nathanielmanistaatgoogle commented on oauth2client/django_util/apps.py:
```diff
@@ -0,0 +1,23 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
```
This whole module looks really really weird; I think a module doc string would explain what it is doing and would justify its existence.

waprin replied:
I agree it's weird, it's mostly to follow a Django convention for which the entire purpose is not totally clear to me, but pretty much all the Django contrib examples I've seen have it. Anyway I added some explanation and linked to the relevant Django docs.

nathanielmanistaatgoogle commented on oauth2client/django_util/decorators.py:
```diff
@@ -0,0 +1,110 @@
+        return curry_wrapper
+
+
+
```
What's up with all these trailing blank lines?

waprin replied:
done in incoming commit (need to figure out how to do automatically do this)

nathanielmanistaatgoogle commented on oauth2client/django_util/signals.py:
```diff
@@ -0,0 +1,18 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
```
Weird module; no module doc string.

waprin replied:
done in incoming commit (the docstring ,kept in a separate module based on Django examples I've seen)

nathanielmanistaatgoogle commented on oauth2client/django_util/views.py:
```diff
@@ -0,0 +1,142 @@
+
+def _get_flow_for_token(csrf_token, request):
+    """ Looks up the flow in session to recover information about requested scopes """
+    flow_pickle = request.session.get(
```
Unnecessary line-wrap?

waprin replied:
done in incoming commit

nathanielmanistaatgoogle commented on oauth2client/django_util/views.py:
```diff
@@ -0,0 +1,142 @@
+    """ Looks up the flow in session to recover information about requested scopes """
+    flow_pickle = request.session.get(
+        _FLOW_KEY.format(csrf_token), None)
+    if flow_pickle is None:
```
How about

```
return None if flow_pickle is None else pickle.loads(flow_pickle)
```

instead of these three lines?

waprin replied:
done in incoming commit

nathanielmanistaatgoogle commented on tests/test_django_orm.py:
```diff
@@ -30,10 +30,11 @@
-django_settings.SECRET_KEY = 'xyzzy'
+os.environ['DJANGO_SETTINGS_MODULE'] = 'tests.test_django_settings'
+from django.conf import settings
+settings.SECRET_KEY = 'xyzzy'
```
How about "this string is not a real Django SECRET_KEY" or something like that?

waprin replied:
done in incoming commit

nathanielmanistaatgoogle commented on tests/test_django_settings.py:
```diff
@@ -0,0 +1,35 @@
+
+import os
+
+SECRET_KEY = 'xyzzy'
```
Same observation here.

waprin replied:
done in incoming commit

nathanielmanistaatgoogle commented on tests/test_django_settings.py:
```diff
@@ -0,0 +1,35 @@
+
+GOOGLE_OAUTH2_CLIENT_ID = 'client_id2'
+GOOGLE_OAUTH2_CLIENT_SECRET = 'hunter2'
+GOOGLE_OAUTH2_SCOPES = ('email',)
```
Can this setting be different than GOOGLE_OAUTH2_DEFAULT_SCOPES defined in **init**.py? If so, make it a different value. If not, express the relationship semantically by defining and using just one constant.

waprin replied:
done in incoming commit

nathanielmanistaatgoogle commented on tests/test_django_settings.py:
```diff
@@ -0,0 +1,35 @@
+GOOGLE_OAUTH2_SCOPES = ('email',)
+
+ROOT_URLCONF = 'tests.test_django_util'
+SECRET_KEY = 'xyzzy'
```
This is the second time you've assigned this name to this value in this file.

waprin replied:
done in incoming commit

nathanielmanistaatgoogle commented:
I've completed a first pass.

waprin commented:
@nathanielmanistaatgoogle thanks for taking time to review. I addressed all your comments (by doing what they asked for). Ready for another pass when you are.

waprin commented:
@nathanielmanistaatgoogle friendly ping this is still waiting.

dhermes commented:
@nathanielmanistaatgoogle PTAL now that you're not OOO (not sure if you'll have time in the shortened week)

waprin commented:
@nathanielmanistaatgoogle It's the season of giving, perhaps you are ready to give this another pass. I know you are tied up in storage PR atm, just your weekly friendly ping this is still here.

nathanielmanistaatgoogle commented:
Thank you for your patience; I am finally getting to this in the next hour.

I apologize if this is slow to swap back into my working memory and I ask the same questions I asked last month.

nathanielmanistaatgoogle commented on oauth2client/django_util/__init__.py:
```diff
@@ -0,0 +1,294 @@
+            settings_instance.GOOGLE_OAUTH2_CLIENT_SECRET
+    else:
+        raise exceptions.ImproperlyConfigured(
+            "Must specify either GOOGLE_OAUTH2_CLIENT_SECRETS_JSON,or both "
```
Missing space after comma on this line?

waprin replied:
done in incoming commit

nathanielmanistaatgoogle commented on oauth2client/django_util/__init__.py:
```diff
@@ -0,0 +1,294 @@
+            settings_instance.GOOGLE_OAUTH2_CLIENT_SECRETS_JSON)
+    elif getattr(settings_instance, "GOOGLE_OAUTH2_CLIENT_ID",
+                 None) and getattr(settings_instance, "GOOGLE_OAUTH2_CLIENT_SECRET", None):
+        return settings_instance.GOOGLE_OAUTH2_CLIENT_ID, \
```
"Do not use backslash line continuation." (https://google.github.io/styleguide/pyguide.html?showone=Line_length#Line_length)

waprin replied:
done in incoming commit

nathanielmanistaatgoogle commented on oauth2client/django_util/__init__.py:
```diff
@@ -0,0 +1,294 @@
+
+def _get_oauth2_client_id_and_secret(settings_instance):
+    """Initializes client id and client secret based on the settings"""
+    if getattr(django.conf.settings, 'GOOGLE_OAUTH2_CLIENT_SECRETS_JSON', None):
```
Avoid making the attribute access twice by binding it locally:

```
secrets_json = getattr(<...>)
if secrets_json is not None:
  return _load_client_secrets(secrets_json)
else:
  <adjective>_id = <expression>
  client_secret = <expression>
  if <adjective>_id is not None and client_secret is not None:
    return <adjective_id>, client_secret
  else:
    raise <...>
```

waprin replied:
done in incoming commit

nathanielmanistaatgoogle commented on oauth2client/django_util/__init__.py:
```diff
@@ -0,0 +1,294 @@
+
+    Attributes:
+      scopes: A list of OAuth2 scopes that the decorators and views will use as defaults
+      request_prefix: The name of the attribute that the decorators use to attach the UserOAuth2
```
Is this line wrapping before eighty columns? GitHub isn't telling me one way or the other.

waprin replied:
I wasn't being careful with column length and wrapping, next commit should be 80 and consistent wrapping.

nathanielmanistaatgoogle commented:
I'm somewhat bothered by the structure of this - rather than a new package with a non-empty `__init__.py` and several other modules, I would sort of expect behavior like this to land in a `django_util.py` module in the `oauth2client` package with other behavior included in package-private helper modules. How much control do we have over how this code is arranged into modules and packages? How strongly is Django pushing the design of this code into its current shape?

waprin commented:
Not sure exactly what you mean by package-private helper modules - I am guessing you mean everything that is part of the public API goes in oauth2client/django_util.py and everything else goes into a helper module or two? I originally did this PR as one file, and it wouldn't be too much work to go back. Django doesn't really force you in any technical sense. However, I laid it out the way I did so it mirrors other Django apps/extensions I looked at:

From the Django project::
https://github.com/django/django/tree/master/django/contrib/admin
https://github.com/django/django/tree/master/django/contrib/auth

In the wild:
https://github.com/ui/django-rq/tree/master/django_rq
https://github.com/omab/django-social-auth/tree/master/social_auth

So it's a choice between being closer to Django style or other extensions in this library (like Flask). @jonparrott suggested I make it closer to Django which made sense to me. Also, keep in mind we are probably going to move this to a contrib package anyway, #326, which also tilts my opinion towards keeping closer to Django style.

But, ultimately I'm not super committed to one way or the other, and if you have something else in mind, I'm fine to change it to whatever you'd like.

nathanielmanistaatgoogle commented:
Yeah, I'm now persuaded that the way this is laid out is just fine.

What's with the failing test result?

Does this need to wait for anything else to go in before it can be merged?

dhermes commented:
Fix the broken RST in the docstring

```
Warning, treated as error:
/home/travis/build/google/oauth2client/oauth2client/django_util/decorators.py:docstring of oauth2client.django_util.decorators.oauth_enabled:27: WARNING: Field list ends without a blank line; unexpected unindent.
```

waprin commented:
@nathanielmanistaatgoogle had to rebase, ready to merge when you are.

nathanielmanistaatgoogle commented on tox.ini:
```diff
@@ -46,12 +47,15 @@ deps = {[testenv]deps}
     python2.6
 commands =
-    nosetests \
+    nosetests  \
```
Erroneously added extra space before backslash?

nathanielmanistaatgoogle commented:
Now that the `contrib` directory exists, should this code be placed there?

jonparrott commented:
Yes, if you've rebased, @waprin, might as well go ahead and move it. Otherwise I can just move it when I move everything else.

waprin commented:
Ok I'll move it now

waprin commented:
Ok, moved to contrib and I did some double manual checking so I think it's good to go. @nathanielmanistaatgoogle

dhermes commented:
Yay!

anthmgoogle commented:
Great to see this feature. Thanks all.
